jax/interpreters/xla.py

```diff
@@ -533,6 +602,52 @@ def _xla_callable(fun: lu.WrappedFun, device, backend, name, *arg_specs):
   else:
     return partial(_execute_replicated, compiled, result_handlers)
 
+def configure_aliasing(
+    c: xc.XlaBuilder,
+    xla_args: Sequence[xe.XlaOp],
+    out_tuple: xe.XlaOp,
+    arg_donate: Sequence[bool],
+    tuple_args: bool,
+):
+  """Configures input/output "must" aliasing based on `arg_donate`."""
+  assert len(xla_args) == len(arg_donate)
+
+  # First for every input array add it to `donations` iff it is a member of
+  # `arg_donate`.
+  donations = defaultdict(deque)
+  for arg_index, arg in enumerate(xla_args):
+    if arg_donate[arg_index]:
+      for param_index, element in flatten_shape(c.GetShape(arg)):
+        key = (element.dimensions(), element.numpy_dtype())
+        if tuple_args:
+          param_number = 0
+          param_index = (arg_index,) + tuple(param_index)
+          donations[key].append((param_number, param_index, arg_index))
+        else:
+          param_number = arg_index
+          donations[key].append((param_number, param_index, arg_index))
+
+  # Consume donations for outputs.
+  for output_index, element in flatten_shape(c.GetShape(out_tuple)):
+    key = (element.dimensions(), element.numpy_dtype())
+    if donations.get(key, ()):
+      param_number, param_index, _ = donations[key].popleft()
+      c.SetUpAlias(output_index, param_number, param_index)
```

Comment:
High-level question: why look for matching outputs instead of having the user specify explicit (input, output) donation pairs? I'm guessing it's more ergonomic to only specify the donated inputs, and maybe more convenient if you don't have specific outputs in mind. The downsides I see are that this could theoretically lead to the wrong matching if there are multiple inputs/outputs with the same shape, and the API is a little subtle since you get an error if there are unmatched inputs (although including the output requirements in the docstrings would probably be sufficient).

(Apologies if I missed some earlier discussion somewhere...)

Reply:
I have a similar question. But presumably the donation API could generalize to user-specified aliasing (pass a dict instead of a set).

Reply:
As per discussion in design review we'll leave this API as is for now (just specifying what buffers you are letting JAX reason about and not telling us specifically what to do with them).